### PR TITLE
Update PX4 Porting Guide organization for clarity

### PR DESCRIPTION
**Pull request by joshuawhitehead**
Attempted to group information in a more consistent (and easily expandable) manner. Updated language in various locations to add clarity and be more consistent within the porting guide and with respect to other documentation files.  Added links and references in a few locations to other PX4 documentation as appropriate.  Corrected multiple dead links.

### CODE REVIEW

File: en/debug/porting-guide.md
Comment (hamishwillee) [inline en/debug/porting-guide.md]: Links within the same doc should be relative. So for example this `https://dev.px4.io/en/middleware/` should be `../middleware/README.md`

File: en/debug/porting-guide.md
Comment (hamishwillee) [inline en/debug/porting-guide.md]: This link will work, as will `(reference-design.md)`. I have a slight preference for the full relative path `(../debug/reference-design.md)` as it reminds me this file needs to be moved to somewhere with a better name!

File: en/debug/porting-guide.md
Comment (hamishwillee) [inline en/debug/porting-guide.md:64]: This is true, though in future these are expected to use same drivers as NuttX

File: en/debug/porting-guide.md
Comment (hamishwillee) [inline en/debug/porting-guide.md]: Calling FMUv5 the current reference config for NuttX based flight controllers feels a little odd because historically the variants are not necessarily sequential - ie FMU versions only really infer binary compatibility and a lower number could be the same level of advanced. I guess it is the most recent and the one we would hope manufacturers would target so leave as is.

**Comment by hamishwillee** [comment]
@joshuawhitehead Yes, this is much much better.

@dagar Is there anything in your recent reorganisational work that invalidates any of this, or which should be captured (in addition)? If not, I say we merge.

File: en/debug/porting-guide.md
Comment (joshuawhitehead) [reply en/debug/porting-guide.md]: Ah, ok, so anything that's in the "Devguide" repo should use a relative link, I should only use a full URL if I need to reference something else, e.g. the User Guide?  I was working through the GitHub interface to make this update and was a little unsure of how it handled things.  I will update this.

File: en/debug/porting-guide.md
Comment (joshuawhitehead) [reply en/debug/porting-guide.md]: Agreed, that makes sense, I will update accordingly.

File: en/debug/porting-guide.md
Comment (joshuawhitehead) [reply en/debug/porting-guide.md:64]: Would it be worth adding a "Future Plans" subsection to add things like that? In the chance you pique someone's interest to help make it happen :-)

File: en/debug/porting-guide.md
Comment (joshuawhitehead) [reply en/debug/porting-guide.md]: Understood about the binary compatibility/feature set, I had read that in other places as well, I had intended it only as you mentioned- If someone is porting to a brand new board, FMUv5 would be the best place to start.  I thought I had seen a similar sentence in the reference-design.md doc, but now I'm not... I will add a note here to remind people that the versions/generations to not correlate with capabilities, simply that v5, being the most recent version, reflects the most up to date methods for hardware and code structure.

**Comment by joshuawhitehead** [comment]
Thank you for all the feedback @hamishwillee !

To add some context: I've been chatting with @dagar on the PX4 slack channel as well, and I am currently working through doing a port for a Cortex-R5 based FMU. My plan is to add as many of my "lessons learned" here as I am able. If you are happy with the changes so far, and are willing to merge them as is, I will wait until that merge happens to make the updates I mentioned in my replies to your comments above.

My plan is to approach the porting guide from the perspective of "Ok, you have a NuttX `hello_world` booting on the board/chip you want to use, now what?", mostly because that's how I'm doing it... If you have any thoughts on a superior approach, I'd be happy to hear them. Some things I would like to see in the guide are:
  * A list of the minimal drivers required for a "working" PX4 port
    * Perhaps even a table showing the capabilities enabled by a particular driver beyond the minimum set as this would allow developers to target their efforts
  * While the cMake files are somewhat "self documenting", a high-level overview of the build process would be nice, particularly to call out any "gotchas" that a developer might encounter, e.g. PX4 does not (currently) use the entire NuttX build system, it only uses NuttX to build some pieces/libraries and does the final linking with its own cMake files
  * ~If you'd like to use a back-up coprocessor (e.g. the STM32F103 on the PixHawk), where does that fit into the whole build process, and what are its requirements?~ (Found this)

Just a few ideas I had at the moment, there will likely be more, and I plan to answer as many of them myself as I am able while working on this port.  Any feedback is welcome, I'm monitoring the PX4 slack channel as well if that would be an easier place to have a conversation.

File: en/debug/porting-guide.md
Comment (dagar) [inline en/debug/porting-guide.md]: This target isn't valid.

File: en/debug/porting-guide.md
Comment (hamishwillee) [reply en/debug/porting-guide.md]: Thanks!

File: en/debug/porting-guide.md
Comment (hamishwillee) [reply en/debug/porting-guide.md]: That is correct. You've done it right throughout pretty much - except for this one.

File: en/debug/porting-guide.md
Comment (hamishwillee) [reply en/debug/porting-guide.md]: I actually wouldn't add a note here (maybe back in the reference-design doc though!). 
What I'd do is change as shown:

FROM
> The following example uses FMUv5 as it is the current [reference configuration](./reference-design.md) for NuttX based flight controllers:

TO

> The following example uses FMUv5 as it is a recent [reference configuration](./reference-design.md) for NuttX based flight controllers:

File: en/debug/porting-guide.md
Comment (hamishwillee) [reply en/debug/porting-guide.md:64]: Not for now. Daniel Agar is going to do this particular task and I don't know of other future plans. If you know of enough to justify them, then definitely.

File: en/debug/porting-guide.md
Comment (hamishwillee) [reply en/debug/porting-guide.md]: @joshuawhitehead Can you fix this up too  - it should just be `make eagle_default` (according to [this](https://dev.px4.io/en/setup/building_px4.html#qurt--snapdragon-based-boards).

**Comment by hamishwillee** [comment]
Hi @joshuawhitehead 

> Thank you for all the feedback @hamishwillee !

You're very welcome. I should be thanking you. I've been trying to get to this document for ages, but 
* I simply do not know enough about porting to take a decent shot at it. 
* Still haven't managed to arrange a discussion with the people who do. 

> If you are happy with the changes so far, and are willing to merge them as is, I will wait until that merge happens to make the updates I mentioned in my replies to your comments above.

Can you please make the minor and specific modifications suggested - I'm then happy to merge (if there are any in the current set of suggestions you think are significant they could wait).

> My plan is to approach the porting guide from the perspective of "Ok, you have a NuttX hello_world booting on the board/chip you want to use, now what?", mostly because that's how I'm doing it... If you have any thoughts on a superior approach, I'd be happy to hear them. 

I think this is a good way to approach the problem - at least you should be able to capture most the issues and build iteratively. At some point you may create enough material that a different organisation is worth considering, but we're a long way from that. 

YOu might start even earlier discussing what sorts of board porting "makes sense" - ie what chips you can use, where to modifications "make sense" (as I said, I haven't had a discussion with a board engineer yet to know what things are doable/reasonable, so this might just be silly).

> Some things I would like to see in the guide are:
> A list of the minimal drivers required for a "working" PX4 port
> Perhaps even a table showing the capabilities enabled by a particular driver beyond the minimum set as this would allow developers to target their efforts

Sounds good. There is lots of work going on with drivers at the moment to capture much of the copy paste used currently into base classes. @dagar is doing most of that, and I presume a lot of it will inject into the device driver docs. 

With respect to driver docs/tables, at the moment we auto-generate module docs for a small subset of [drivers](https://dev.px4.io/en/middleware/modules_driver.html) - a little more on this in https://dev.px4.io/en/apps/module_template.html

What I would like to do is autogenerate more complex user  documentation about what is supported (from source). For example, at the moment this is handcrafted: https://docs.px4.io/en/sensor/rangefinders.html - but we could generate it from in source docs using only minor script enhancements of the module documentation. Anyway, just FYI - on my list as drivers get updated. 

> While the cMake files are somewhat "self documenting", a high-level overview of the build process would be nice, particularly to call out any "gotchas" that a developer might encounter, e.g. PX4 does not (currently) use the entire NuttX build system, it only uses NuttX to build some pieces/libraries and does the final linking with its own cMake files

Could not agree more. The current docs assume you already know how the bits fall together. 

> If you'd like to use a back-up coprocessor (e.g. the STM32F103 on the PixHawk), where does that fit into the whole build process, and what are its requirements? 

Yes. So that to me is part of that "what can you do/what should you do/what options are there" discussion that is currently missing. 

Thanks again.

File: en/debug/porting-guide.md
Comment (joshuawhitehead) [reply en/debug/porting-guide.md]: Will do.  This was copy-paste from the current version of the porting guide, so the target may have changed since that was originally written.

**Review by hamishwillee — APPROVED** [review]

**Comment by hamishwillee** [comment]
Thanks @joshuawhitehead - a clear improvement over previous text.

**Pull request merged**
